adr: merge tendermint/spec repository into tendermint/tendermint (#7775)
tendermint/tendermint: docs/architecture/adr-076-combine-spec-repo.md

# ADR 076: Combine Spec and Tendermint Repositories 

## Changelog

- 2022-02-04: Initial Draft. (@tychoish)

## Status

Accepted.

## Context

> While the specification for tendermint was original in the same

Reviewer: ```suggestion
While the specification for Tendermint was original in the same
```

repository as the Go implementation, at some point the specification
was split from the core tendermint repository and maintained
separately from the implementation. While this makes sense in
promoting a conceptual separation of specification and implementation,

> in practice this separation was a premature optimization apparently

Reviewer: Do we have context from the original split? I'm pretty curious as to what justification was given for this in the past to make sure we're not ripping up a chesterton's fence.

Reviewer 2: I tried to find an issue or commit logs to explain it, but I wasn't able to find a clear explanation. Even the commits at the time of the split are mostly just "do the split". If we _do_ have a doc somewhere, I agree it would be nice to link it.

Author: yeah, this is the main reason that I made an ADR for this even though I think this might be a bit of a process overshoot.

> While the specification for tendermint was original in the same
> repository as the Go implementation, at some point the specification
> was split from the core tendermint repository and maintained
> separately from the implementation. While this makes sense in
> promoting a conceptual separation of specification and implementation,
> in practice this separation was a premature optimization apparently
> aimed at supporting alternate implementations of tendermint. 

Reviewer 2: Trivial lexical nits.

```suggestion
While the specification for Tendermint was originally in the same
repository as the Go implementation, at some point the specification
was split from the core repository and maintained separately from the
implementation. While this makes sense in promoting a conceptual
separation of specification and implementation, in practice this
separation was a premature optimization, apparently aimed at supporting
alternate implementations of Tendermint. 
```

> However, maintaining a spec repo has ended up being a premature
> optimization: there are no current projects to develop alternate
> implementations of tendermint based on the common specification, and
> having separate repositories creates an ongoing maintenance burden to
> support versioning and publication.

Reviewer 2: Some suggestions to consolidate themes, and to remove duplication of the previous paragraph's conclusion.

```suggestion
The operational and documentary burden of maintaining a separate
spec repo has not returned value to justify its cost. There are no active
projects to develop alternate implementations of Tendermint based on the
common specification, and having separate repositories creates an ongoing
burden to coordinate versions, documentation, and releases.
```

## Decision

> The specification repository should be merged

Reviewer 2: ```suggestion
The specification repository will be merged back into the Tendermint core repository.
```

> Stakeholders including representatives from the maintainers of the
> spec, the go implementation, and the tendermint rust library, agreed
> to merge the repositories in the tendermint core dev meeting on 27
> January 2022.

Reviewer 2: ```suggestion
Stakeholders including representatives from the maintainers of the
spec, the Go implementation, and the Tendermint Rust library, agreed
to merge the repositories in the Tendermint core dev meeting on 27
January 2022.
```

Reviewer 2: Maybe we should log who participated in that decision for the record?

## Alternative Approaches

> The main alternative to this was *do nothing* and continue to maintain
> separate versioning for both repositories, complicating the release
> process for both repositories and creating a difficult to resolve 

Reviewer 2: Suggestions to clarify that "do nothing" was relative to merging, not other changes, and finish up an incomplete

```suggestion
The main alternative we considered was to keep separate repositories,
and to introduce a coordinated versioning scheme between the two, so
that users could figure out which spec versions go with which versions
of the core implementation.

We decided against this on the grounds that it would further complicate
the release process for _both_ repositories, without mitigating any of
the other existing issues.
```

## Detailed Design

> Within the `tendermint` repository, do the following to merge the
> repositories keeping history: 
> 
> ```bash
> git remote add spec git@github.com:tendermint/spec.git
> git fetch spec
> git checkout -b spec-master spec/master
> mkdir spec
> git ls-tree -z --name-only HEAD | xargs -0 -I {} git mv {} subdir/
> git commit -m "spec: organize specification prior to merge"
> git checkout -b spec-merge-mainline origin/master
> git merge --allow-unrelated-histories spec-master
> ```

Reviewer 2: Suggestion to summarize what the script is doing at a high level.

```suggestion
Clone and merge the master branch of the `tendermint/spec` repository
as a branch of the `tendermint/tendermint`, to ensure the commit history
of both repositories remains intact.

### Implementation Instructions

1. Within the `tendermint` repository, execute the following commands 
   to add a new branch with the history of the master branch of `spec`:

```bash
git remote add spec git@github.com:tendermint/spec.git
git fetch spec
git checkout -b spec-master spec/master
mkdir spec
git ls-tree -z --name-only HEAD | xargs -0 -I {} git mv {} subdir/
git commit -m "spec: organize specification prior to merge"
git checkout -b spec-merge-mainline origin/master
git merge --allow-unrelated-histories spec-master
```
```

> This branch should be merged into the `tendermint/tendermint`
> repository as a normal branch. This commit can also be backported to
> the 0.35 branch, if needed.

Reviewer 2: ```suggestion
This merges the spec into the `tendermint/tendermint` repository as a normal
branch. This commit can also be backported to the 0.35 branch, if needed.
```

We might wind up wanting parts but not all of it, depending on when we do this merge. The state of the protos in particular is somewhat dissynchronized.

> Then, make a final change to the specification repository adding a
> redirect to the README and marking the repository as archived. All
> issues should be recreated in the `tendermint/tendermint` repository.

Reviewer 2: ```suggestion
2. In the specification repository, add redirect to the README and mark
   the repository as archived. All issues should be recreated in the
   `tendermint/tendermint` repository.
```

Is it possible we can move the issues? They're all within the same organization, so I think we could transfer them. It might require a bit of scripting, but I think should be tractable (we had to do this for Kythe at one point when we split out the org).

> issues should be recreated in the `tendermint/tendermint` repository.

Reviewer: I believe we can transfer the issues across: https://docs.github.com/en/issues/tracking-your-work-with-issues/transferring-an-issue-to-another-repository 

This will maintain the links so that documents referencing the issues etc. won't lose these useful info.

Author: nice! lets do that.

## Consequences

### Positive

> Easier maintenance for the specification will obviate a number of
> complicated and annoying versioning problem, and will help prevent the
> possibility where the specification drifts from the implementation.

Reviewer 2: ```suggestion
Easier maintenance for the specification will obviate a number of
complicated and annoying versioning problems, and will help prevent the
possibility of the specification and the implementation drifting apart.
```

> Additionally, collocating the specification will help encourage
> cross-polenation  and collaboration between engineers focusing on the
> specification and the protocol and engineers focusing on the
> implementation.

Reviewer 2: ```suggestion
Additionally, co-locating the specification will help encourage
cross-pollination and collaboration, between engineers focusing on the
specification and the protocol and engineers focusing on the implementation.
```

### Negative

> Co-locating the spec and go implementation has the potential effect of
> prioritizing the Go implementation with regards to the spec, and
> making it difficult to think about alternate implementations of the
> tendermint algorithm. Although we may want to foster additional
> tendermint implementations in the future, this isn't a going concern
> in our roadmap at the moment, and *not* merging these repos doesn't
> change the fact that the Go implementation of tendermint is already the
> primary implementation.

Reviewer 2: ```suggestion
Co-locating the spec and Go implementation has the potential effect of
prioritizing the Go implementation with regards to the spec, and
making it difficult to think about alternate implementations of the
Tendermint algorithm. Although we may want to foster additional
Tendermint implementations in the future, this isn't an active goal
in our current roadmap, and *not* merging these repos doesn't
change the fact that the Go implementation of Tendermint is already the
primary implementation.
```
